### DISCUSSION TITLE
How can I squash multiple commits into a single commit before merging?

### DISCUSSION THREAD

**Discussion by Ketan2120** [Ketan2120/Nexus-project-2-Restaurant-website#1] [Q&A] [answered]
How can I squash multiple commits into a single commit before merging?

**Comment by ketan270** [comment, marked as answer]
To squash multiple commits into a single commit:

Ensure you are on the branch with the commits you want to squash:
Copy code
git checkout feature-branch
Start an interactive rebase for the last n commits (replace n with the number of commits you want to squash):
css
Copy code
git rebase -i HEAD~n
In the editor that opens, change the word pick to squash (or s) for the commits you want to combine into the one above.
Save and close the editor. Another editor will open to combine the commit messages.
Save and close the editor to complete the rebase.
Q: How can I resolve merge conflicts in a rebased branch?

A: To resolve merge conflicts during a rebase:

During the rebase, Git will pause and highlight the conflicts. Resolve the conflicts in the affected files.
After resolving the conflicts, stage the resolved files:
csharp
Copy code
git add resolved-file
Continue the rebase process:
kotlin
Copy code
git rebase --continue
Repeat the process until the rebase completes.
GitHub Actions and Automation
Q: How can I create a custom GitHub Action?

A: To create a custom GitHub Action:

Create a new directory in your repository called .github/actions/my-action.
In this directory, create an action.yml file with the following content:
yaml
Copy code
name: 'My Custom Action'
description: 'An example custom action'
inputs:
  my-input:
    description: 'An input for my custom action'
    required: true
runs:
  using: 'node12'
  main: 'index.js'
Add your action code in index.js:
javascript
Copy code
const core = require('@actions/core');

try {
  const myInput = core.getInput('my-input');
  console.log(`Hello ${myInput}`);
} catch (error) {
  core.setFailed(error.message);
}
Use your custom action in a workflow:
yaml
Copy code
name: Custom Action Workflow

on: [push]

jobs:
  my_job:
    runs-on: ubuntu-latest
    steps:
    - name: Checkout code
      uses: actions/checkout@v2
    - name: Run custom action
      uses: ./.github/actions/my-action
      with:
        my-input: 'World'
Q: How can I trigger a GitHub Action only on specific paths?

A: To trigger a GitHub Action only on changes to specific paths:

Define the paths in your workflow trigger:
yaml
Copy code
name: Path Specific Workflow

on:
  push:
    paths:
      - 'src/**'
      - '!src/exclude-this-file.txt'

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - name: Checkout code
      uses: actions/checkout@v2
    - name: Run build
      run: npm run build
Repository Security and Compliance
Q: How can I enforce code review rules in my GitHub repository?

A: To enforce code review rules:

Go to your repository settings and navigate to "Branches."
Add a branch protection rule for your main branch.
Configure the following settings:
Require pull request reviews before merging.
Specify the number of required reviewers.
Enable "Require status checks to pass before merging."
Optionally, enable "Require signed commits."
Q: How can I use secret scanning to protect sensitive data in my repository?

A: To use GitHub secret scanning:

Enable secret scanning in your repository settings under "Security & analysis."
GitHub will automatically scan for sensitive data like API keys, tokens, and credentials.
If sensitive data is detected, GitHub will notify you with details on how to address the issue.
Collaborative Development Strategies
Q: How can I use GitHub Codespaces for collaborative development?

A: To use GitHub Codespaces:

Ensure Codespaces is enabled for your repository.
Click the "Code" button and then the "Codespaces" tab.
Create a new Codespace or open an existing one.
Develop and collaborate in a fully configured, cloud-based development environment.
Q: How can I manage project milestones and issues effectively on GitHub?

A: To manage project milestones and issues:

Create milestones for major project goals under the "Issues" tab.
Assign issues to the appropriate milestones.
Use labels to categorize and prioritize issues.
Regularly review and update the progress of milestones and issues.
Advanced CI/CD Pipelines
Q: How can I implement a multi-step CI/CD pipeline with GitHub Actions?

A: To implement a multi-step CI/CD pipeline:

Create a workflow file, e.g., ci-cd-pipeline.yml.
Define the workflow with multiple jobs and dependencies:
yaml
Copy code
name: Multi-step CI/CD Pipeline

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - name: Checkout code
      uses: actions/checkout@v2
    - name: Set up Node.js
      uses: actions/setup-node@v2
      with:
        node-version: '14'
    - name: Install dependencies
      run: npm install
    - name: Run tests
      run: npm test

  deploy:
    runs-on: ubuntu-latest
    needs: build
    steps:
    - name: Checkout code
      uses: actions/checkout@v2
    - name: Deploy to production
      run: npm run deploy
      env:
        DEPLOY_KEY: ${{ secrets.DEPLOY_KEY }}
